[coordinates.featurizer] Add selector for non-symmetric heavy atoms

maxentile opened the pull request:
Follows: https://github.com/choderalab/automatic-state-decomposition/blob/ffbec93383c01db1a0cd62ec9eaae48d93f24c2f/decompose-f95/src/decompose.f90#L278-L287

maxentile committed:
[coordinates.featurizer] Add selector for non-symmetric heavy atoms

Follows: https://github.com/choderalab/automatic-state-decomposition/blob/ffbec93383c01db1a0cd62ec9eaae48d93f24c2f/decompose-f95/src/decompose.f90#L278-L287

jchodera commented:
Thanks, @maxentile!

Can you add a corresponding test [here](https://github.com/markovmodel/PyEMMA/blob/8e2502ef48226977d40e5053ba132d2fc26c6451/pyemma/coordinates/tests/test_featurizer.py#L129-L130)?

maxentile committed:
[coordinates.featurizer] Split up selection

maxentile committed:
[coordinates.featurizer] Test select_non_symmetry_related_heavy_atoms

jchodera commented:
Anything else we should do here before this is merged, @marscher?

marscher commented:
Fine from my side. @gph82 what do you think?

gph82 commented:
Honestly my first thought was to try to include this in mdtraj directly, which does most of the atom selection for us (backbone, residues Cas etc)....makes more sense to me. Perhaps create an extended "select" submodule of us in the future. Until then, what about putting the selection string in pyemma/utils somewhere (and keep the call to "select" in featurizer.py)

jchodera commented:
Since this is working now, what about a three-step plan?
- Merge this as-is
- Add to mdtraj
- Refactor to use mdtraj after next mdtraj release

marscher commented:
I'm fine with that.

franknoe commented:
- Is there a shorter name than `...non_symmetry_related..`?
- Is the first exclude line excluding all H's or only H's being called "H"?
- If you exclude all H's (which you should if you want heavy atoms), what's the function of the last line?
  `exclusions.append("(resname == ARG and name == NH1) or (resname == ARG and name == NH2)")`

franknoe commented:
... on the first point: should be easy to find things using autocomplete, and have related things close so I suggest `select_heavy_nsym`. 

Alternatively you could just add a parameter to the existing `select_heavy` function `symmetry_related=True`. I think this is the better option because it pools functionality and avoids combinatorial explosion of number functions due to different parameter combos.

maxentile commented:
> Is the first exclude line excluding all H's or only H's being called "H"?

Ahh, good catch! I meant to exclude all H's -- better to exclude "mass < 2", or, even better, add the other exclusions to `select_Heavy`, as you suggest.

> If you exclude all H's (which you should if you want heavy atoms), what's the function of the last line? `exclusions.append("(resname == ARG and name == NH1) or (resname == ARG and name == NH2)")`

I believe NH1 and NH2 are nitrogens, not hydrogens.

> Alternatively you could just add a parameter to the existing `select_heavy` function `symmetry_related=True`. I think this is the better option because it pools functionality and avoids combinatorial explosion of number functions due to different parameter combos.

Agreed! I'll move this into `select_Heavy`. I think `symmetry_related=True` might be ambiguous (would you interpret "True" as "included" or "excluded" ?) -- is `exclude_symmetry_related=False` okay?

maxentile committed:
[coordinates.featurizer] Move selector for non-symmetric heavy atoms

- Moves the exclusion of symmetry-related heavy atoms into `select_Heavy` as an option
- Also excludes all hydrogens, not just those named "H"

maxentile committed:
[coordinates.featurizer] Fix optional exclusion

- Use correct name of option

maxentile committed:
[coordinates.featurizer] Test exclude_symmetry_related

jchodera commented:
>  is exclude_symmetry_related=False okay?

I like that!

franknoe commented:
yes, that's fine.

Am 23/08/16 um 06:42 schrieb John Chodera:

> ```
> is exclude_symmetry_related=False okay?
> ```
> 
> I like that!
> 
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub 
> https://github.com/markovmodel/PyEMMA/pull/898#issuecomment-241625390, 
> or mute the thread 
> https://github.com/notifications/unsubscribe-auth/AGMeQk20g5aMEnmVD_P9cr--Z30k3ES2ks5qinpTgaJpZM4Jjid7.

## 

---

Prof. Dr. Frank Noe
Head of Computational Molecular Biology group
Freie Universitaet Berlin

Phone: (+49) (0)30 838 75354
Web:   research.franknoe.de

## Mail:  Arnimallee 6, 14195 Berlin, Germany

franknoe commented:
can be merged when passing.

maxentile commented:
Sorry, I'm having trouble understanding the current test failures.

In Travis, [tests pass in 5 of 6 environments](https://travis-ci.org/markovmodel/PyEMMA/builds/154349637). The error occurs in `test_bayesian_msm.py:test_transition_matrix_samples` in the Mac OS, Python 3.5, Numpy 1.11 environment:
https://travis-ci.org/markovmodel/PyEMMA/jobs/154349642#L2155-L2163 .

In Appveyor, [tests pass in 3 of 4 environments](https://ci.appveyor.com/project/marscher/pyemma/build/1.0.2271). The error occurs in `test_its_bmsm` in the Python 2.7, ARCH=64 environment: https://ci.appveyor.com/project/marscher/pyemma/build/1.0.2271/job/1278g6ygk183iebk#L1912

I'm confused because [test_bayesian_msm.py](https://github.com/markovmodel/PyEMMA/blob/devel/pyemma/msm/tests/test_bayesian_msm.py) doesn't appear to touch the coordinates module.

Any idea how I should address this?

jchodera commented:
I believe those are stochastic test failures.

marscher commented:
yes. Nothing related to your changes! Thank you!
